clang/lib/AST/Decl.cpp

```diff
@@ -4507,9 +4507,14 @@ bool FieldDecl::isZeroSize(const ASTContext &Ctx) const {
 
   // Otherwise, [...] the circumstances under which the object has zero size
   // are implementation-defined.
-  // FIXME: This might be Itanium ABI specific; we don't yet know what the MS
-  // ABI will do.
-  return true;
+  if (!Ctx.getTargetInfo().getCXXABI().isMicrosoft())
+    return true;
+
+  // MS ABI: has nonzero size if it is a class type with class type fields,
+  // whether or not they have nonzero size
+  return !llvm::any_of(CXXRD->fields(), [](const FieldDecl *Field) {
+    return Field->getType()->getAs<RecordType>();
```

Comment:
My point here was, the isEmpty test above means we already know that CXXRD has no non-static data members of class type: https://clang.llvm.org/doxygen/classclang_1_1CXXRecordDecl.html#ac570e2716d2cb020851575aa4e5fafe0

So this test doesn't seem to do anything.

Reply:
Oh, I see. The comment above the isEmpty check specifies "-- has subobjects of nonzero size or bit-fields of nonzero length", so I was assuming that isEmpty is taking the zero size stuff into account. I found it mentioned in this comment as well https://github.com/llvm/llvm-project/blob/004a7cea705d63441307a5e0429b6786d49042c2/clang/include/clang/AST/CXXRecordDeclDefinitionBits.def#L49. 

Also, in the last commit I added tests to verify that this check is working.